### PR TITLE
Create a base CLI class for simplifying scripts (HPC-7528)

### PR DESCRIPTION
**Pull request by itkovian**
- [x] code
- [x] tests

### CODE REVIEW

File: lib/vsc/utils/script_tools.py [hunk 1]
```diff
@@ -210,3 +226,179 @@ def critical_exception_handler(self, tp, value, traceback):
...
+        If errors evals to false, and this is not a dry_run (or forced group or accountn selection),
+        it is considered success and creates the cache file with current time
+        """
+        if current_time is None:
+            current_time = self.current_time
+
+        if errors:
+            logging.warning("Could not process all %s", errors)
+            self.warning("Not all processed")
+        elif not (self.options.dry_run or self.options.groups or self.options.accounts):
```
Comment (wpoely86) [inline lib/vsc/utils/script_tools.py]: what's groups and accounts in this context?

File: lib/vsc/utils/script_tools.py [hunk 2]
```diff
@@ -210,3 +226,179 @@ def critical_exception_handler(self, tp, value, traceback):
...
+        sys.exit(NAGIOS_EXIT_CRITICAL[0])
+
+    def critical_exception(self, msg, exception):
+        """
+        Convenience method: report exception and critical method
+        """
+        logging.exception("%s: %s", msg, exception)
+        self.critical("%s (%s)" % (msg, type(exception)), log=False)
+
+    def do(self, dry_run):  #pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (itkovian) [reply lib/vsc/utils/script_tools.py]: It's passed in the `main`.

File: lib/vsc/utils/script_tools.py [hunk 3]
```diff
@@ -210,3 +226,178 @@ def critical_exception_handler(self, tp, value, traceback):
         message = "Script failure: %s - %s" % (tp, value)
         sys.exc_clear()
         self.critical(message)
+
+
+class CLI(object):
+    """
+    Base class to implement sync clients.
```
Comment (wdpypere) [inline lib/vsc/utils/script_tools.py]: CLI means command line interface to me, it seems like a bad name for something about sync clients.

File: lib/vsc/utils/script_tools.py [hunk 4]
```diff
@@ -210,3 +226,178 @@ def critical_exception_handler(self, tp, value, traceback):
...
+                    mtime = '%s_%s' % (mtime, convert_to_unix_timestamp(self.current_time))
+                kwargs['modified_since'] = mtime
+
+    def warning(self, msg):
+        """
+        Convenience method that calls ExtendedSimpleOptions critical and exists with nagios warning exitcode
+        """
+        logging.warning(msg)
+        self.fulloptions.critical(msg)
+        sys.exit(NAGIOS_EXIT_WARNING[0])
```
Comment (wdpypere) [inline lib/vsc/utils/script_tools.py]: why not  use nagios exit with errocode  https://github.com/hpcugent/vsc-utils/blob/509c43c9a5e06d23fb00c83b8af24cd0377ea0f0/lib/vsc/utils/nagios.py#L137
then you can drop these functions?

or use a simplenagios object which has these functions?

Comment (itkovian) [reply lib/vsc/utils/script_tools.py]: @stdweird ^ Any reason?

**Comment by wdpypere** [comment]
@itkovian I actually thought you would rename the CLI class to something else. :smile:  does this mean we will be using this class for all our CLI's?

**Comment by itkovian** [comment]
I remain open to naming proposals.

File: test/script_tools.py [hunk 5]
```diff
@@ -88,3 +89,60 @@ def test_threshold_custom_setting(self, mock_proceed, mock_lock, mock_lockfile):
...
+            'dry_run': False,
+            'ha': None,
+            'help': None,
+            'ignoreconfigfiles': None,
+            'info': False,
+            'locking_filename': '/var/lock/setup.lock',
+            'nagios_check_filename': '/var/cache/setup.nagios.json.gz',
+            'nagios_check_interval_threshold': 0,
+            'nagios_report': False,
+            'nagios_user': 'nagios',
```
Comment (wdpypere) [inline test/script_tools.py:128]: can you set the default to 'nrpe' plz?

File: test/script_tools.py [hunk 6]
```diff
@@ -88,3 +89,60 @@ def test_threshold_custom_setting(self, mock_proceed, mock_lock, mock_lockfile):
...
+            'debug': False,
+            'disable_locking': False,
+            'dry_run': False,
+            'ha': None,
+            'help': None,
+            'ignoreconfigfiles': None,
+            'info': False,
+            'locking_filename': '/var/lock/setup.lock',
+            'nagios_check_filename': '/var/cache/setup.nagios.json.gz',
+            'nagios_check_interval_threshold': 0,
```
Comment (wdpypere) [inline test/script_tools.py:126]: I never understood the purpose of this one.

Comment (itkovian) [reply test/script_tools.py:126]: It's hazy for me as well, but it had something to do with rechecks from the monitoring?

Comment (wdpypere) [reply test/script_tools.py:126]: what would the script need to know about that?

Comment (wdpypere) [reply test/script_tools.py:126]: icinga itself schedules and reschedules according to its config. how would that affect the script? (and how is it kept in sync with what icinga schedules?)

Comment (itkovian) [reply test/script_tools.py:126]: All right, but this PR is not the place to change this, I think.

Comment (wdpypere) [reply test/script_tools.py:126]: but if it is not needed, why define it?

Comment (itkovian) [reply test/script_tools.py:126]: nagios.py has
~~~
       @param threshold: Seconds to determines how old the gzipped JSON data may be
                         before reporting an unknown result. This can be used to check if the script that uses the
                         reporter has run the last time and succeeded in writing the cache data. If the threshold <= 0,
                         this feature is not used.
~~~

Comment (wdpypere) [reply test/script_tools.py:126]: ach, so the name is just odd and confusing. (and only needed in corner cases.)

File: lib/vsc/utils/script_tools.py [hunk 7]
```diff
@@ -36,13 +36,29 @@
...
 from vsc.utils.timestamp_pid_lockfile import TimestampedPidLockfile
 
+DEFAULT_TIMESTAMP = "20140101000000Z"
+TIMESTAMP_FILE_OPTION = 'timestamp_file'
+DEFAULT_CLI_OPTIONS = {
+    'start_timestamp': ("The timestamp form which to start, otherwise use the cached value", None, "store", None),
+    TIMESTAMP_FILE_OPTION: ("Location to cache the start timestamp", None, "store", None),
+}
+MAX_DELTA = 3
+MAX_RTT = 2 * MAX_DELTA + 1
```
Comment (wdpypere) [inline lib/vsc/utils/script_tools.py:60]: what is RTT? Regie voor Telegraaf en Telefoon? :wink:

Comment (itkovian) [reply lib/vsc/utils/script_tools.py:60]: Don't ask me, I took it from the Sync class code, it's for determining is some time is out of sync.

**Comment by boegel** [comment]
@itkovian This needs a sync with `master` after the merge of #67 (which was needed to avoid issues elsewhere, in `vsc-modules` and `vsc-rpms`)

File: lib/vsc/utils/script_tools.py (hunk 7 above)
Comment (stdweird) [reply lib/vsc/utils/script_tools.py:60]: round trip time

File: lib/vsc/utils/script_tools.py [hunk 8]
```diff
@@ -210,3 +226,173 @@ def critical_exception_handler(self, tp, value, traceback):
...
+            options[TIMESTAMP_FILE_OPTION] = tuple(tsopt)
+
+        options.update(self.CLI_OPTIONS)
+
+        if TIMESTAMP_FILE_OPTION not in options and self.TIMESTAMP_MANDATORY:
+            raise Exception("no mandatory %s option defined" % (TIMESTAMP_FILE_OPTION,))
+
+        return ExtendedSimpleOption(options)
+
+    def _update_kwargs_modified_since(self, kwargs):
```
Comment (stdweird) [inline lib/vsc/utils/script_tools.py:284]: this method is probably only useful in teh AP client

File: lib/vsc/utils/script_tools.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stdweird) [reply lib/vsc/utils/script_tools.py:314]: the idea was/is that we make this explicit, typically the do method should handle the dry run. at the very least, the `do` should do a `log.error(nothing implemented)` or something

File: lib/vsc/utils/script_tools.py [hunk 9]
```diff
@@ -210,3 +226,173 @@ def critical_exception_handler(self, tp, value, traceback):
...
+    def make_time(self):
+        """
+        Get start time (from commandline or cache), return current time
+        """
+        try:
+            (start_timestamp, current_time) = retrieve_timestamp_with_default(
+                getattr(self.options, TIMESTAMP_FILE_OPTION),
+                start_timestamp=self.options.start_timestamp,
+                default_timestamp=DEFAULT_TIMESTAMP,
+                delta=-MAX_RTT,  # make the default delta explicit, current_time = now - MAX_RTT seconds
```
Comment (stdweird) [inline lib/vsc/utils/script_tools.py:336]: so the delat can probably be 0 if there is not external system to compare to. in the AP it has to be > 0, but not for everything. (obviously, won't matter much, it's only a few seconds anyway, maybe make this a bit clear somewhere)

File: lib/vsc/utils/script_tools.py (hunk 8 above)
Comment (itkovian) [reply lib/vsc/utils/script_tools.py:284]: Or any other thing that interacts with something?
